app/jobs/maintenance_tasks/task_job.rb

```diff
@@ -25,20 +25,21 @@ def retry_on(*, **)
 
     private
 
+    EnumerationContext = Struct.new(:cursor, keyword_init: true)
+
     def build_enumerator(_run, cursor:)
-      cursor ||= @run.cursor
-      collection = @task.collection
-
-      case collection
-      when ActiveRecord::Relation
-        enumerator_builder.active_record_on_records(collection, cursor: cursor)
-      when Array
-        enumerator_builder.build_array_enumerator(collection, cursor: cursor)
-      when CSV
-        JobIteration::CsvEnumerator.new(collection).rows(cursor: cursor)
+      context = EnumerationContext.new(
+        cursor: cursor || @run.cursor,
+      )
+
+      if @task.throttle_condition
+        JobIteration::EnumeratorBuilder.new(self).build_throttle_enumerator(
+          @task.enumerator_builder.enumerator(context: context),
+          throttle_on: @task.throttle_condition,
+          backoff: @task.throttle_backoff
+        )
       else
-        raise ArgumentError, "#{@task.class.name}#collection must be either "\
-          'an Active Record Relation, Array, or CSV.'
+        @task.enumerator_builder.enumerator(context: context)
```

Comment:
We could extract the wrapping, that way you wouldn't need to repeat the initial enumerator creation.

```ruby
enumerator = @task.enumerator_builder.enumerator(context: context)

if @task.throttle_condition
  enumerator = JobIteration::EnumeratorBuilder.new(self).build_throttle_enumerator(
    enumerator,
    throttle_on: @task.throttle_condition,
    backoff: @task.throttle_backoff,
  )
end

enumerator
```

Reply:
Yes, definitely!